addons/website_sale/models/website.py

```diff
@@ -284,10 +284,14 @@ def get_current_pricelist(self):
         return pricelist
 
     def sale_product_domain(self):
-        return expression.AND([self._product_domain(), self.get_current_website().website_domain()])
+        return expression.AND([
+            self._product_domain(),
+            self.get_current_website().website_domain(),
+            self.get_current_website().company_domain()
+        ])
 
     def _product_domain(self):
-        return [('sale_ok', '=', True)]
+        return [('sale_ok', '=', True), ('website_published', '=', True)]
```

Comment:
For the reviewer, could you confirm this modification can be done or it's a wanted behavior to take care of unpublished product in filter (even if not logged in) ?

Reply:
Just passing by but generally, for all models having `website_published` field, this check is supposed to be appended in every `search` request automatically through `ir.rule`, see `domain_force`.
Isn't it the case here?

Reply:
As filter use the value of it an unpublished product, I want to say no the case for filter/search part
https://github.com/odoo/odoo/blob/16.0/addons/website_sale/controllers/main.py#L173

Seems ok for the product list part
https://github.com/odoo/odoo/blob/16.0/addons/website_sale/security/website_sale.xml#L4

I can use this "published" domain update only in search, it seemed logical to me to align list and filter on shop but maybe there is a better way to handle this case
![image](https://user-images.githubusercontent.com/55689427/230608256-bf347b48-b218-451d-aa39-958aa8a744bb.png)

Reply:
3 tests failing due to `('website_published', '=', True)` addition, no idea to make filter use ir.rule too
```
TestUi.test_02_admin_shop_custom_attribute_value_tour
TestWebsiteSaleImage.test_01_admin_shop_zoom_tour
TestUi.test_01_admin_shop_custom_attribute_value_tour
```

1 test failing due to ??? (fail on branch 16.0 too on my side)
```
TestUi.test_01_free_delivery_when_exceed_threshold
```

Reply:
As said by RDE, website_published is automatically managed by an ir.rule (and the company filtering should be too).

Could you provide clear steps to reproduce the bug or even better add a test covering the problematic flow ?

There are two potential causes for the problem you have AFAICS:
* either a sudo disables the ir rules and must be investigated
* ir rules are bypassed somehow (which seems to be the case for the price range filter) when they shouldn't be:

https://github.com/odoo/odoo/blob/e34e908a06c18ecee5d762f6fca785b23977f47d/addons/website_sale/controllers/main.py#L390-L396

Reply:
On runbot:
- Set a price of a product to 999999 (easy visual identification) and update it to "Unpublished"
--> Website filter (price range) use unpublished product for filter data computing
- Back to the product, set its company to a company different than the website
--> Website filter (price range) use other company product for filter data computing (published or not)

[Live demo](https://drive.google.com/file/d/1_H6BhDCjk7CT0kyGZbe4ocNfaZjLX1tJ/view?usp=sharing)


Domain used in the sql request that you mention is: 
```
['&', ('sale_ok', '=', True), ('website_id', 'in', (False, 1))]
```
With the fix
```
['&', '&', '&', ('sale_ok', '=', True), ('website_published', '=', True), ('website_id', 'in', (False, 1)), ('company_id', 'in', (False, 1))]
```

It's why I do the modification to get published product and company product only, but if ir rules must be manage there too, idk how so I take any infos to do the right fix.

Reply:
pushed a fixup commit that seems to fix the problem, following the standard logic of the main `_search` methods of the ORM.
Take a look and feel free to squash the commits afterwards (or tell me if there's still a bug).

Reply:
Good to know this method exists, it works, thx